File: LinkedIn-backend/src/main/java/com/ex/service/PostService.java
package com.ex.service;

import com.ex.dao.HibDao;
import com.ex.models.Category;
import com.ex.models.Post;
import com.ex.models.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Set;

/**
 * This class will be used as a medium in between the controller and the DAO
 * It will cover any business logic and invoke methods to obtain data from the DAO
 *
 */
@Service
public class PostService {

    private HibDao dao;

    @Autowired
    public PostService(HibDao dao){
        this.dao = dao;
    }

    /**
     * This method is used to get all of the posts from the database
     *
     * @return - a list of all the posts in  the database
     */
    public List<Post> getAllPosts(){
        return dao.getAllPosts();
    }

    /**
     * This method will get the post that has the id of the parameter
     *
     * @param id - the od of the post this is fetching
     * @return - the post object related to the id
     */
    public Post getPost(int id){
        return dao.getPostById(id);
    }

    /**
     * This method will get all posts that were posted  by a specific user
     *
     * @param username - the username of the user to get the posts of
     * @return - a list of posts that the specific user posted
     */
    public List<Post> getPosts(String username) {

        User user = dao.getExistingUser(username);

        return dao.getPostsByUser(user);

    }

    /**
     * This will get all the posts that a specific user applied to
     *
     * @param username - the username of the user to get the applied posts from
     * @return - a set of posts that the user has applied to
     */
    public Set<Post> getAppliedPosts(String username) {

        User user = dao.getExistingUser(username);

        return dao.getPostsByApplied(user);
    }

    /**
     * This will get all posts for the specified category
     *
     * @param id - the id of the category that it want the posts of
     * @return - a list of posts that have the specified category
     */
    public List<Post> getPostsByCat(int id) {

        Category c = dao.getCategoryById(id);

        return dao.getPostsByCategory(c);
    }

    /**
     * This will delete a post
     *
     * @param id - the id of the post that is being deleted
     * @return - true if it was deleted, false if it wasn't
     */
    public boolean deletePost(int id){
        Post post = dao.getPostById(id);

        if(dao.deletePost(post) >= 1){
            return true;
        }
        else{
            return false;
        }
    }
}

File: angular-linkedin/src/app/services/post.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Post } from '../models/post';
import { User } from '../models/user';

@Injectable({
  providedIn: 'root'
})
/**
 * This is used to get any post data from our backend
 */
export class PostService {

  constructor(private http:HttpClient) { }

  /**
   * This will return an Observable that holds an array of all the posts in our database
   */
  getAllPosts(): Observable<Post[]>{
    return this.http.get<Post[]>("http://localhost:8080/LinkedIn_backend_war_exploded/api/posts");
  }

  /**
   * This will return an observable that holds an array of all the posts of a certain user
   * @param u - the user to get the posts of
   */
  getPostsForUser(u:User): Observable<Post[]>{
    return this.http.get<Post[]>("http://localhost:8080/LinkedIn_backend_war_exploded/api/posts/poster/"+u.id);
  }

  /**
   * This will return an observable that holds a string that will say "deleted" if deleted and "not-deleted" if it wasn't
   * @param p - the post to delete
   */
  deletePost(p:Post): Observable<string>{
    return this.http.get<string>("http://localhost:8080/LinkedIn_backend_war_exploded/api/posts/delete/" + p.id);
  }

  /**
   * This will return an observable that holds an array of posts that a user applied to
   * @param u - the user to get the applied post of
   */
  appliedPosts(u:User): Observable<Post[]>{
    return this.http.get<Post[]>("http://localhost:8080/LinkedIn_backend_war_exploded/api/posts/applied/"+u.username);
  }

}
